hashicorp/terraform-provider-tfe
# Add ability to upload Policies in a PolicySet resource

## Description

This PR augments the `tfe_policy_set` resource to allow uploading of local policies. This is done primarily through the use of [PolicySetVersions](https://www.terraform.io/docs/cloud/api/policy-sets.html#upload-policy-set-versions). 

**Example usage**

This functionality only works when there is no VCS repo attached to a PolicySet. The usage is:

```hcl
data "tfe_slug" "policy" {
  // path to local policies
  source_path = "policies/my-policy-set"
}

resource "tfe_policy_set" "test" {
  name          = "my-policy-set"
  description   = "A brand new policy set"
  organization  = "<org-name>"

  // reference to the tfe_slug data source
  slug = data.tfe_slug.policy
}
```

Behind the scenes, this creates a new PolicySetVersion and then Uploads the contents of the `source_path` in the `tfe_slug` data source.

## Testing plan

1. Pull down this branch and `make build` the provider. Make sure to enable [development override for the provider](https://www.terraform.io/docs/cli/config/config-file.html#development-overrides-for-provider-developers).
2. Create a new directory and `terraform init`
3. Copy the `tfe/test-fixtures/policy-set-version` directory to your local working directory under `policies/my-policy-set`
4. Add this to your config

```
data "tfe_slug" "policy" {
  source_path = "policies/my-policy-set"
}

resource "tfe_policy_set" "test" {
  name          = "my-policy-set"
  description   = "A brand new policy set"
  organization  = tfe_organization.test.name
  slug = data.tfe_slug.policy
}

output "policy" {
  value = tfe_policy_set.test
}

output "data_policy" {
  value = data.tfe_slug.policy
}
```

5. Run `terraform apply`
6. Validate in the UI in this organization that the policy set is created, and verify the terraform output that the `status` is `ready`. You can also [curl to get the PolicySetVersion and confirm the status](https://www.terraform.io/docs/cloud/api/policy-sets.html#show-a-policy-set-version).
7. Then edit/add a new file in the `policies/my-policy-set` directory and run `terraform apply` and notice an PolicySet update taking place (terraform outputs 1 changed). 

## Documentation

Images of the new documentation is put as a PR comment near the file.



## External links

- [API documentation](https://www.terraform.io/docs/cloud/api/policy-sets.html#upload-policy-set-versions)
- [go-tfe PR](https://github.com/hashicorp/go-tfe/pull/230)

Closes #279 

## Output from acceptance tests

```
terraform-provider-tfe % TFE_TOKEN=<token> TFE_HOSTNAME=tfe-zone-4c292d5b.ngrok.io TF_ACC=1 TF_LOG_PROVIDER=DEBUG  gotestsum --format testname -- ./tfe/... -v -run TestAccTFEPolicySet_version
PASS tfe.TestAccTFEPolicySet_version (7.01s)
PASS tfe.TestAccTFEPolicySet_version_updateCreate (12.06s)
PASS tfe.TestAccTFEPolicySet_version_no_vcs_repo (1.20s)
PASS tfe

DONE 3 tests in 26.078s
```
